pmd-java/src/main/resources/category/java/performance.xml

```diff
@@ -112,6 +164,184 @@ public class Test {
         </example>
     </rule>
 
+    <rule name="AvoidCalendarDateCreation"
+          since="6.17.0"
+          language="java"
+          message="A Calendar is used to create a Date or DateTime, this is expensive."
+          class="net.sourceforge.pmd.lang.rule.XPathRule"
+          typeResolution="true"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_performance.html#avoidcalendardatecreation">
+        <description>Problem: A Calendar is a heavyweight object and expensive to create. &#13;
+            Solution: Use 'new Date()', Java 8+ java.time.[Local/Zoned]DateTime.now() or joda time '[Local]DateTime.now()'.
+        </description>
+        <priority>2</priority>
+        <properties>
+            <property name="version" value="1.0"/>
+            <property name="xpath">
+                <value><![CDATA[
+//PrimaryPrefix[Name[ends-with(@Image, 'Calendar.getInstance')]] [count(../PrimarySuffix) > 2 and ../PrimarySuffix[last()-1][@Image = 'getTime' or @Image='getTimeInMillis']]
+|
+//Block/BlockStatement//Expression/PrimaryExpression/
+PrimaryPrefix/Name[typeIs('java.util.Calendar') and (ends-with(@Image,'.getTime') or ends-with(@Image,'.getTimeInMillis'))]
+|
+//ClassOrInterfaceType[typeIs('org.joda.time.DateTime') or typeIs('org.joda.time.LocalDateTime')][../Arguments/ArgumentList/Expression/PrimaryExpression/PrimaryPrefix/Name[ends-with(@Image, 'Calendar.getInstance')]]
+	         ]]></value>
+            </property>
+        </properties>
+        <example>
+            <![CDATA[
+public class DateStuff {
+    private Date bad1() {
+        return Calendar.getInstance().getTime(); // now
+    }
+    private Date good1a() {
+        return new Date(); // now
+    }
+    private LocalDateTime good1b() {
+        return LocalDateTime.now();
+    }
+    private long bad2() {
+        return Calendar.getInstance().getTimeInMillis();
+    }
+    private long good2() {
+        return System.currentTimeMillis();
+    }
+}
+            ]]>
+        </example>
+    </rule>
+
+    <rule name="AvoidConcatInAppend"
```

Comment:
Possibly. We don't have that rule in the subset we use, probably because we didn't update that selection for some time. There is certainly overlap, not sure if they are exactly the same.

Reply:
On a separate PR for each rule: Juan asked me to create a PR per batch, a couple of rules together. Not sure of best choice here.

Reply:
They appear to have the same intention, but there are some differences in their implementation:

For example 
* your rule flags `wrappedLine.append(str + "c", a, c);` correctly whereas the other rule gives up
* your rule flag `sb.append(1 + 2)`, which is not a concatenation, so that's a false positive
* your rule considers concatenation with a non-static final field as a constant expression, which it isn't
* the other rule doesn't flag any cases of `new StringBuilder().append(...)`, ie use of append directly on the constructor invocation, whereas yours does it correctly

I think we should fix the false negatives of the existing rule but not merge this new rule.

Reply:
I agree, we should enhance the existing rule InefficientStringBuffering instead of adding this new rule.

FWIW: I've noticed, that the found violation don't highlight the append - at least for the [3 violations found in checkstyle](https://chunk.io/pmd/473fc5bbeb07453d8a603380175858ed/diff/checkstyle/index.html): I wouldn't know, what to fix, if I get these violations. We should aim to report exactly the node which contains the concatenation. Maybe the reason for this is the `//`, that Clément mentioned?